alloc/src/chain.rs

```diff
@@ -288,10 +288,9 @@ impl<A: Allocator + Clone> Drop for ChainAllocator<A> {
                     // moved to the stack before the chunk is dropped, so it's
                     // alive and valid after the chunk is dropped below.
                     chain_node_ptr = unsafe {
-                        core::ptr::addr_of!((*non_null.as_ptr()).prev)
-                            .read()
-                            .get()
-                            .read()
+                        // Save to variable to avoid a dangling temporary.
+                        let unsafe_cell = core::ptr::addr_of!((*non_null.as_ptr()).prev).read();
+                        unsafe_cell.get().read()
                     };
```

Comment:
Wow, very interesting catch. With my knowledge from other programing languages and things that compilers can do :tm: I'd fully expect the old version of the code to be equivalent to the new one, but I guess Rust has other ideas ;) 

On the other hand, as long as the compiler points out the sharp edge, I think I'm fine with Rust being a bit quirky on this :D

Reply:
This could be a false positive. It's probably worth time to think about it and report upstream if necessary, except that this is already a tangent from my main work. But we're firmly in unsafe code here, and unsafe Rust can be subtle, so it's certainly possible that this is an issue. I believe miri is running on this code--hence the skips of failures for miri as I double-checked--and it didn't trigger anything. So... not sure who is right.

Reply:
Reading the error message and https://doc.rust-lang.org/reference/destructors.html#drop-scopes I'm leaning towards this being a real issue.

It looks like destruction order and scopes are very precisely defined, rather than the "if the compiler sees you're not using it, it may just clean it up much earlier" than other languages I'm used to.

So by introducing the local variable, it looks like the compiler will be forced to keep `unsafe_cell` alive until the end of the scope it's defined on, even though at first sight it looked like it wouldn't need to.

(I guess this means that if the Rust compiler wants to optimize further it'll need to prove that the earlier drop can happen in a safe way, which it probably won't even try in the presence of `unsafe`)